lib/Zef/Client.pm6

```diff
@@ -106,12 +106,36 @@ class Zef::Client {
                     phase   => BEFORE,
                     message => "Dependencies: {@specs-batch.map(*.name).unique.join(', ')}",
                 });
-                next unless my @needed = @specs-batch\               # The current set of specs
+
+                next unless my @orig-needed = @specs-batch\               # The current set of specs
                     .grep({ not @skip.first(*.contains-spec($_)) })\ # Dists in @skip are not needed
                     .grep(-> $spec { not @!exclude.first({ $_.spec-matcher($spec) }) })\
                     .grep(-> $spec { not @!ignore.first({ $_.spec-matcher($spec) }) })\
                     .grep({ $skip-installed ?? self.is-installed($_).not !! True });
-                my @identities = @needed.map(*.identity);
+
+                my @prereq-candidates;
+                my @needed;
+                my @identities = gather for @orig-needed -> $needed {
+                    my @candidates;
+                    if $needed.isa(Zef::Distribution::DependencySpecification::Any)
+                        and $needed.specs.first({
+                            @candidates = self.find-candidates(:$upgrade, $_.identity);
+                            try @candidates.append: self.find-prereq-candidates(:$upgrade, @candidates) if @candidates;
```

Comment:
I'm trying to think of a way to make this non-recursive, which is something I went out of my way to avoid in this method for reasons I can't remember (probably ease of parallelization). Unfortunately it looks like things such as passing around string identities instead of objects in some of the original code might make that a little harder.